apps/prairielearn/src/ee/pages/studentCourseInstanceUpgrade/studentCourseInstanceUpgrade.ts

```diff
@@ -17,7 +44,221 @@ router.get(
+          prairielearn_user_id: user.user_id,
+        },
+        line_items: lineItems,
+        mode: 'payment',
+        success_url: `${urlBase}/success?session_id={CHECKOUT_SESSION_ID}`,
+        cancel_url: urlBase,
+      });
+
+      await insertStripeCheckoutSessionForUserInCourseInstance({
+        session_id: session.id,
+        institution_id: institution.id,
+        course_instance_id: course_instance.id,
+        user_id: user.user_id,
+        data: session,
+        plan_names: planNames,
+      });
+
+      if (!session.url) throw error.make(500, 'Stripe session URL not found');
+
+      res.redirect(session.url);
+    } else {
+      throw error.make(400, `Unknown action: ${req.body.__action}`);
+    }
+  }),
+);
+
+router.get(
+  '/success',
+  asyncHandler(async (req, res) => {
+    const institution = InstitutionSchema.parse(res.locals.institution);
+    const course = CourseSchema.parse(res.locals.course);
+    const course_instance = CourseInstanceSchema.parse(res.locals.course_instance);
+
+    if (!req.query.session_id) throw error.make(400, 'Missing session_id');
+
+    const stripeSessionId = z.string().parse(req.query.session_id);
+
+    const localSession = await getStripeCheckoutSessionBySessionId(stripeSessionId);
+    if (!localSession) {
+      throw new Error(`Unknown Stripe session: ${stripeSessionId}`);
+    }
+
+    const stripe = getStripeClient();
+    const session = await stripe.checkout.sessions.retrieve(stripeSessionId);
+
+    // Verify that the session is associated with the current course instance
+    // and user. We shouldn't hit this during normal operations, but an attacker
+    // could try to replay a session ID from a different course instance or user.
+    if (
+      localSession.institution_id !== institution.id ||
+      localSession.course_instance_id !== course_instance.id ||
+      localSession.user_id !== res.locals.authn_user.user_id
+    ) {
+      throw error.make(400, 'Invalid session');
+    }
+
+    if (session.payment_status === 'paid') {
+      if (!localSession.plan_grants_created) {
+        // Create plan grants and mark the session as completed.
+        //
+        // Doing these mutations in a GET handler isn't great, but we have
+        // reasonable protection in place against replay attacks, and it would
+        // be difficult to perform a CSRF attack because the session must have
+        // been created in Stripe and must refer to the same user and course instance.
+        await runInTransactionAsync(async () => {
+          for (const planName of localSession.plan_names) {
+            await ensurePlanGrant({
+              plan_grant: {
+                plan_name: planName,
+                type: 'stripe',
+                institution_id: institution.id,
+                course_instance_id: course_instance.id,
+                user_id: res.locals.authn_user.id,
+              },
+              authn_user_id: res.locals.authn_user.id,
+            });
```

Comment:
The use of `ensurePlanGrant` here means that on the off-chance we end up in a state where the user already had the plan grant *but* it was of a different type, we won't update the type to `stripe` in this case. We generally guard against this, but there's a race condition of sorts where a user could start a stripe session, be granted a plan manually, and then complete the purchase.

This seems exceedingly unlikely to happen in practice; just wanted to call it out in case we wanted to handle this differently (e.g. changing the `type` of the existing grant).

Reply:
Discussed with @nwalters512 and we decided against making any change here.